### PR TITLE
Provide code fix for S125: remove commented out code

### PR DESCRIPTION
**Pull request by Corniel**
As mentioned at #313, it would be nice to have a code fix that removes comment out code.

### CODE REVIEW

**Comment by Corniel** [comment]
Note that the test fails, although I copied the output from Visual Studio that is the actual result. It could be considered to remove some extra new-line trivia, but I did not yet try to include those who could be removed.

**Comment by pavel-mikula-sonarsource** [comment]
> Note that the test fails, although I copied the output from Visual Studio that is the actual result. It could be considered to remove some extra new-line trivia, but I did not yet try to include those who could be removed.

CodeFixes are sometimes tricky - updating it to "actual" it might break it again. When reading the error message, pay attention to the "because" message - should be one of these

>VerifyWhileDocumentChanges updates the document until all issues are fixed, even if the fix itself creates a new issue again

>VerifyFixAllProvider runs Xxxxx once

Typically, if the "fixed" version can still raise the issue, the 1st one will appear. In such case, the CodeFix needs to be further improved.

**Comment by gregory-paidis-sonarsource** [comment]
Hey @Corniel , just to add some more information on what Pavel suggested:
We are validating the codefixes in two steps:
- [One change at a time](https://github.com/SonarSource/sonar-dotnet/blob/71ccb2b464127e8924a1b885be7f09590158804b/analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/CodeFixVerifier.cs#L45)
- [All the changes at once](https://github.com/SonarSource/sonar-dotnet/blob/71ccb2b464127e8924a1b885be7f09590158804b/analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/CodeFixVerifier.cs#L69)

These two validations surprisingly do not always yield the same results.

The first one, applies the codefix in a while-loop, until the document no longer mutates.
The second one, applies the codefix _once_ in every single location it was reported on.

This means that it is possible for the "expected" to be different between these two validations, in which case you will probably need to make the codefix implementation a bit tighter. 

I suggest to try setting up a breakpoint inside your codefix on a trivial case, (this is the one I tried):
```csharp
// Noncompliant@+1 :
// ;
using System;
```

...and then see the behavior of the code. For example, while I was debugging it, I noticed that the location of the comment you are trying to track was **not** the same between `VerifyWhileDocumentChanges` and `VerifyFixAllProvider` runs:

```csharp
// this gave me a different location/span between the 2 "verify" methods.
var comment = diagnostic.Location.SourceSpan; 
```

Also, pay attention to how the comments marked as `Noncompliant` are transformed into `Fixed` comments.
Since comments are used to assert the testcase, rules about trivia are slightly more annoying, since comments play two roles (raising diagnostics from the rule, raising assertions from the verifiers), which is not usually the case.

**Review by gregory-paidis-sonarsource — CHANGES_REQUESTED** [review]
Nice work, I left a comment about the breaking tests.

**Comment by Corniel** [comment]
> Hey @Corniel , just to add some more information on what Pavel suggested: We are validating the codefixes in two steps:

I have reduce the fix case to just one simple example:

``` C#
class Compliant
{
    int SingleLine() { return 42; } // Single line comment.
    int BlockComment() { return 42; } /* Single line comment. */
}

class Noncompliant
{
    // Noncompliant@+1
    int SingleLine() { return 42; }// return 17;
}
```

That revolves in:

``` C#
class Compliant
{
    int SingleLine() { return 42; } // Single line comment.
    int BlockComment() { return 42; } /* Single line comment. */
}

class Noncompliant
{
    // Noncompliant@+1
    int SingleLine() { return 42; }// return 17;
}
```

That goes right for one by one, and fails for all at once. If I inspect the root node, both cases seem to return the same code, but still as you already pointed out, the location reported on is different. I don't see why that would be case, if the root node is still the same one, and only one issue is reported.

**Comment by gregory-paidis-sonarsource** [comment]
> That goes right for one by one, and fails for all at once. If I inspect the root node, both cases seem to return the same code, but still as you already pointed out, the location reported on is different. I don't see why that would be case, if the root node is still the same one, and only one issue is reported.

@Corniel it is very possible that this is a bug in how Roslyn resolves trivia on one-by-one versus all-at-once scenarios. 
It is quite unfortunate because it has nothing to do with this PR, but it cannot be merged until this discreptancy is dealt with. SyntaxTrivia is always more annoying than the rest of the AST. I actually debugged your code and IMO it seems reasonable, I can't explain the difference in the reported location. 

I feel like what needs to be done if one had the time to spend on this, is to try to reproduce it on the latest Roslyn version, and it if it still incosistent make a ticket on the roslyn repo, probably about [FixAllContent](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.codefixes.fixallcontext?view=roslyn-dotnet-4.3.0).

I will also close this PR, since even if the Roslyn team fixes the issue, we are backwards compatible with older Roslyn versions and the codefix will be consistent in latest VS version, but incosistent in all the previous ones.

**Comment by Corniel** [comment]
That would be too bad. I like the rule, and this fix would make it way easier to use.

**Comment by gregory-paidis-sonarsource** [comment]
> That would be too bad. I like the rule, and this fix would make it way easier to use.

Hey again @Corniel, sorry for the delay :) 
I think what you could do is disable the `FixAllProvider` and add a bunch more testcases. I discussed it a bit internally and I don't think we should care about `FixAll` here, since the user should go with case-by-case basis on each issue. 

To do that, you need to override the Provider in your codefix like this:

```csharp
public override FixAllProvider GetFixAllProvider() => null;
```
...and the UTs should only fire the one-by-one fix provider, instead of both.

If you still wanna work on this, please revert to the previous implementation (i think you changed it to showcase the weird behavior of code fix providers?), add a lot more testcases and ask for review again :)

**Comment by Corniel** [comment]
@gregory-paidis-sonarsource I've extended the testcases (and coverage). Block comments do not work yet (I did not have time yet to see why), but I thought it would be already worth telling I made some (I think) nice progress. I remove lines that only contained commented out code, and obsolete trailing whitespace.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                    ClearEmptiedLines(trailing, token, comment);
+                    RemoveWhitespace(trailing, token);
+                    newNode = node.WithTrailingTrivia(trailing.Values);
+                }
+                else if (Remove(node.GetLeadingTrivia(), comment) is { } leading)
+                {
+                    ClearEmptiedLines(leading, node.ChildTokens().First(), comment);
+                    newNode = node.WithLeadingTrivia(leading.Values);
+                }
+                else
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:59]: Why do you have this else block?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        var spans = trivias.ToDictionary(x => x.GetLocation().GetLineSpan(), x => x);
+        return spans.Remove(comment) ? spans : null;
+    }
+
+    private static void ClearEmptiedLines(Trivias triviaList, SyntaxToken token, FileLinePositionSpan comment)
+    {
+        if (!ShareLine(comment, token))
+        {
+            foreach (var shared in triviaList.Keys.Where(x => ShareLine(comment, x)))
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:81]: Here you are editing the collection you are enumerating inside the loop, which is an `InvalidOperation`.
Consider a local variable like:

```csharp
var redundantNewLines = triviaList.Keys.Where(x => ShareLine(comment, x)).ToList(); // ToList is necessary, to avoid deferred execution
```

...and enumerate that in the for-loop.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            {
+                triviaList.Remove(shared);
+            }
+        }
+    }
+
+    private static void RemoveWhitespace(Trivias triviaList, SyntaxToken token)
+    {
+        foreach (var whitespace in triviaList
+            .Where(x => x.Value.IsKind(SyntaxKind.WhitespaceTrivia) && ShareLine(x.Key, token))
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:91]: Same comment as the above, make a local variable for the enumerated LINQ query and evaluate it eagerly via `ToList`.

**Review by gregory-paidis-sonarsource — CHANGES_REQUESTED** [review]
Looks better, but you need to take into account all the possible comment-types and their usecases:

- SingleLine: `//`
- MultiLine: `/* */`
- ~~SingleLineDocumentation `///`~~
- ~~MultiLineDocumentation `/** */`~~

The rule is raising only for the first two, so you don't have to think about the documentation variants.

About the usecases, there are some more you should consider:

```csharp
readonly int case1 = 42; /* random comment
                      * return 42;
                      */

readonly int case2 = /* 
         * return 42;
         */
    42;

int /* return 42; */ Case3() => 42;
```

These are quite tricky to parse. For example, case3 is three trivia in a row:
- Whitespace
- MultiLineComment
- Whitespace

There, you should delete only one of the two whitespaces.
Also, pay attention to _where_ exactly the rule raises.
For example, in case1, the rule raises on the line that contains the code `returns 42;`, even though the comment spans 3 lines. This means you will have imperfect information about the comment intersecting with the token it is attached to, line-wise.

To visualize it a bit better:

```csharp

readonly int x = /* this day is 
         * return 42;
//^^^^^^^^^^^^^^^^^^^
         * very beautiful */
```

...even though this is a single multi-line comment trivia.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 1 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:59]: Well, the good question is, what to do when you get there, because that can happen. So you could argue it is currently there for debug purposes.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 2 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:81]: The funny thing is, it does not. I'm not sure why, as the iterator of the KeyCollection does a version check. Anyhow, I've added `.ToArray()` (just a bit cheaper then `.ToList()`).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 1 above)
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:59]: I think you should refactor this so that there is no else block. 
Either return inside the if and else-if branches, or always return the same thing, with `node==newNode` on the "useless" branch, so that the tree is the same.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 2 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:81]: I did...

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 1 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: @gregory-paidis-sonarsource 

I've totally re-implemented the code fix.

It is now based on the token, instead of the node, as it turns out that not all trivia ar bound (leading or trailing) to any node.

Furthermore, I've decided that I'll only remove multi line comments if all lines are code (or whitespace, or `/*` or `*/`).

**Comment by Corniel** [comment]
@gregory-paidis-sonarsource The PR says you require changes. Is that still the case?

**Comment by gregory-paidis-sonarsource** [comment]
> @gregory-paidis-sonarsource The PR says you require changes. Is that still the case?

hey @Corniel , I will take a look soon and let you know, sorry for the delay :)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        private SyntaxToken NewTrailing()
+        {
+            var trailing = ShareLine(Token, Comment)
+                ? Trailing.Where(t
+                    => !ShareLine(t, Comment)
+                    || (t.IsKind(SyntaxKind.EndOfLineTrivia) && ShareLine(t, Comment))).ToList()
+                : Trailing;
+
+            // this can happen for multi line comments within an expression.
+            if (trailing.Count == 0 && ShareLine(Token.GetNextToken(), Comment))
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:102]: Optional: What do you think of adding a concrete example here?

```csharp
            // this can happen for multi line comments within an expression.
            // int /* return 42; */ x;
            //    ^                ^    one of the spaces is needed

```

I think it's a great catch, but I could not reason about it before running the UTs and stepping into it.
(It's kind of ironic that to add this comment, you would need to add an issue of the same rule :) )

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        public bool IsTrailing { get; }
+        public SyntaxTrivia Comment { get; }
+        public List<SyntaxTrivia> Leading { get; }
+        public List<SyntaxTrivia> Trailing { get; }
+        public SyntaxToken Token => Comment.Token;
+
+        public Task<Document> ChangeDocument(Document document, SyntaxNode root) =>
+            Task.FromResult(document.WithSyntaxRoot(root.ReplaceToken(Token, NewToken)));
+
+        public SyntaxToken NewToken =>
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:90]: Could this be a method? Something does not feel right with it being a property.
This is the part of your code that does all the clever stuff to figure out how to re-construct the SyntaxTrivia.

Maybe `GenerateNewToken()`?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            IsTrailing ? NewTrailing() : NewLeading();
+
+        private SyntaxToken NewTrailing()
+        {
+            var trailing = ShareLine(Token, Comment)
+                ? Trailing.Where(t
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:96]: Align it like this, as per the [style guide](https://github.com/SonarSource/sonar-dotnet/blob/master/docs/coding-style.md#multi-line-statements):

```csharp
? Trailing.Where(t => !ShareLine(t, Comment)
                      || (t.IsKind(SyntaxKind.EndOfLineTrivia) && ShareLine(t, Comment))).ToList()

```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            if (trailing.Count == 0 && ShareLine(Token.GetNextToken(), Comment))
+            {
+                trailing.Add(SyntaxFactory.Space);
+            }
+
+            return Token
+                .WithLeadingTrivia(Leading)
+                .WithTrailingTrivia(trailing);
+        }
+        private SyntaxToken NewLeading() => Token
```
Comment (gregory-paidis-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:111]: Move `Token` to the next line please

**Review by gregory-paidis-sonarsource — CHANGES_REQUESTED** [review]
Looks very good, great job!

Left suggestions mostly about:
- Add more UTs
- Styling/structural nitpicks

If you apply all the fixes and the new UTs are green, I think this is mergable 🥇

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    private static bool IsCode(SyntaxTrivia trivia) =>
+        trivia.GetLineNumbers().Count() == 1
+        || Array.TrueForAll(Lines(trivia), IsCode);
+
+    private static bool IsCode(string line)
+    {
+        var trimmed = line.Trim();
+        return trimmed == "/*"
+            || trimmed == "*/"
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:57]: I added this one. Good case!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+            IsTrailing = Trailing.Remove(comment);
+
+            if (!IsTrailing)
+            {
+                Leading.Remove(comment);
+            }
+        }
+
+        public bool IsTrailing { get; }
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:82]: I made both the class and the properties private. That is indeed perfectly fine.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        public SyntaxTrivia Comment { get; }
+        public List<SyntaxTrivia> Leading { get; }
+        public List<SyntaxTrivia> Trailing { get; }
+        public SyntaxToken Token => Comment.Token;
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:86]: I changed it, although to me, using the expression here tells the story better: the token is the token of the comment.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 4 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:103]: That case is covered with: `int /* return 17; */ WithinLine() { return 42; }`.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 5 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:91]: Made it a method (and private).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 6 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:97]: Moved the Where constraint to a method. (I really don't like this style guide by the way.)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 4 above)
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:103]: I meant to add it actually in the code, yes. 
On the one hand, I think it would be nice for someone reading this code for the first time.
On the other hand, it's a code smell :)

I will leave it up to you on this one.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CommentedOutCode.SingleLine.ToFix.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
+﻿// return "Some code at the start of a file".Trim();
+// Leading single line comment.
+class Fixes
```
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CommentedOutCode.SingleLine.ToFix.cs:3]: From the perspective of the code fix provider:
``` C#
 int x =
            // return 42;
            42;
```

Is not different than: `int x = 42; // return 42`, so adding an extra test is not increasing the coverage, and to me, also not adding to the story. That is the reason I opted for creating separate test cases in the first place.

As far as I understood: `int codeBefore; // return 42; ~~~ this is a comment` is not different then `int codeBefore; // return 42;`. If that is not the case, then the fix does not take all required logic into action, and has te be adjusted.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 4 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs:101]: I already added it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 6 above)
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: I see you refactored it into a `XOR`. Semantically, this is not the same operation. 

If you make the [Map](https://en.wikipedia.org/wiki/Karnaugh_map) of the operations, you can see this:

c1 = `!ShareLine`
c2 = `IsEndOfLineTrivia`

|  c1 | c2  | c1 + c1!*c2  | c1 XOR c2  |  identical |
|---|---|---|---|---|
| 0  | 0  |  0  | 0  | ✅   |
| 0  | 1  |  1 |  1 | ✅   |
| 1  | 0  |  1 |  1 | ✅   |
| 1  | 1  |  1 |  0 | ❗   |

Basically, `before` you were keeping the trivia that "did not share a line **AND** were `EndOfLine`", but `after` the refactor, you do not.

Is this intended?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: I don't see the change 🤔 
I pulled it locally as well, maybe you forgot to push it?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 6 above)
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: Note: it is not my favorite styling guide either, but I like first-class methods-as-arguments in `Where` clauses like you did more anyways 😄

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CommentedOutCode.SingleLine.ToFix.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CommentedOutCode.SingleLine.ToFix.cs:3]: Fair points. You convinced me, no need to alter the logic of the codefix. 
Resolving this.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: I rolled it back. It turns out that is brakes the behavior.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 6 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: It was an XOR: !a || (a && b) => !a ^ b

Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: I do not think these two expressions are the same.

Let's say you take the case of `(!a, b)` being `true`:

- `!a || (a && b)` => `true`
- `!a ^ b` => `false`

Am I missing something?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: what part was breaking?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 6 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: No, you're not. It was late. I wasn't thinking straight. Anyway, it should be an OR, not an XOR. that being said, the fact that changing from OR to XOR did not make any test fail, suggests we miss a scenario.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: Empty lines where removed. I had no time to investigate why. But with using the expression again, it worked.

[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: It's okay, leave it that way. :)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs (hunk 6 above)
Comment (gregory-paidis-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CommentedOutCodeCodeFix.cs]: That's possible. Or it's something that can never happen, based on the way the AST is structured.
I think the simplified OR is pretty readable.

**Review by gregory-paidis-sonarsource — APPROVED** [review]
LGTM!
